{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species_name</th>\n",
       "      <th>genome_file_name</th>\n",
       "      <th>cds_file_name</th>\n",
       "      <th>original_name</th>\n",
       "      <th>cds_url</th>\n",
       "      <th>genome_url</th>\n",
       "      <th>proteome_url</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>saccharomyces_cerevisiae</td>\n",
       "      <td>saccharomyces_cerevisiae_genomic.fna</td>\n",
       "      <td>saccharomyces_cerevisiae_cds.fna</td>\n",
       "      <td>saccharomyces_cerevisiae</td>\n",
       "      <td>curl -H \"Accept: application/zip\" \"https://api...</td>\n",
       "      <td>curl -H \"Accept: application/zip\" \"https://api...</td>\n",
       "      <td>curl -H \"Accept: application/zip\" \"https://api...</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fusarium_oxysporum</td>\n",
       "      <td>fusarium_oxysporum_genomic.fna</td>\n",
       "      <td>fusarium_oxysporum_cds.fna</td>\n",
       "      <td>fusarium_oxysporum</td>\n",
       "      <td>curl -H \"Accept: application/zip\" \"https://api...</td>\n",
       "      <td>curl -H \"Accept: application/zip\" \"https://api...</td>\n",
       "      <td>curl -H \"Accept: application/zip\" \"https://api...</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>candida_albicans</td>\n",
       "      <td>candida_albicans_genomic.fna</td>\n",
       "      <td>candida_albicans_cds.fna</td>\n",
       "      <td>candida_albicans</td>\n",
       "      <td>curl -H \"Accept: application/zip\" \"https://api...</td>\n",
       "      <td>curl -H \"Accept: application/zip\" \"https://api...</td>\n",
       "      <td>curl -H \"Accept: application/zip\" \"https://api...</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cryptococcus_neoformans</td>\n",
       "      <td>cryptococcus_neoformans_genomic.fna</td>\n",
       "      <td>cryptococcus_neoformans_cds.fna</td>\n",
       "      <td>cryptococcus_neoformans</td>\n",
       "      <td>curl -H \"Accept: application/zip\" \"https://api...</td>\n",
       "      <td>curl -H \"Accept: application/zip\" \"https://api...</td>\n",
       "      <td>curl -H \"Accept: application/zip\" \"https://api...</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>saccharomyces_kudriavzevii</td>\n",
       "      <td>saccharomyces_kudriavzevii_genomic.fna</td>\n",
       "      <td>saccharomyces_kudriavzevii_cds.fna</td>\n",
       "      <td>saccharomyces_kudriavzevii</td>\n",
       "      <td>curl -H \"Accept: application/zip\" \"https://api...</td>\n",
       "      <td>curl -H \"Accept: application/zip\" \"https://api...</td>\n",
       "      <td>curl -H \"Accept: application/zip\" \"https://api...</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2429</th>\n",
       "      <td>achaetomium_macrosporum</td>\n",
       "      <td>achaetomium_macrosporum_cbs_532_94_v1_0_genomi...</td>\n",
       "      <td>achaetomium_macrosporum_cbs_532_94_v1_0_cds.fna</td>\n",
       "      <td>achaetomium_macrosporum_cbs_532_94_v1_0</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>MycoCosm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2430</th>\n",
       "      <td>acarospora_strigata</td>\n",
       "      <td>acarospora_strigata_cbs_132363_genomic.fna</td>\n",
       "      <td>acarospora_strigata_cbs_132363_cds.fna</td>\n",
       "      <td>acarospora_strigata_cbs_132363</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>MycoCosm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2431</th>\n",
       "      <td>absidia_padenii</td>\n",
       "      <td>absidia_padenii_nrrl_2977_v1_0_genomic.fna</td>\n",
       "      <td>absidia_padenii_nrrl_2977_v1_0_cds.fna</td>\n",
       "      <td>absidia_padenii_nrrl_2977_v1_0</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>MycoCosm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2432</th>\n",
       "      <td>absidia_caerulea</td>\n",
       "      <td>absidia_caerulea_nrrl1315_v1_0_genomic.fna</td>\n",
       "      <td>absidia_caerulea_nrrl1315_v1_0_cds.fna</td>\n",
       "      <td>absidia_caerulea_nrrl1315_v1_0</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>MycoCosm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2433</th>\n",
       "      <td>abortiporus_biennis</td>\n",
       "      <td>abortiporus_biennis_cirm_brfm_1778_v1_0_genomi...</td>\n",
       "      <td>abortiporus_biennis_cirm_brfm_1778_v1_0_cds.fna</td>\n",
       "      <td>abortiporus_biennis_cirm_brfm_1778_v1_0</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>MycoCosm</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2434 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    species_name  \\\n",
       "0       saccharomyces_cerevisiae   \n",
       "1             fusarium_oxysporum   \n",
       "2               candida_albicans   \n",
       "3        cryptococcus_neoformans   \n",
       "4     saccharomyces_kudriavzevii   \n",
       "...                          ...   \n",
       "2429     achaetomium_macrosporum   \n",
       "2430         acarospora_strigata   \n",
       "2431             absidia_padenii   \n",
       "2432            absidia_caerulea   \n",
       "2433         abortiporus_biennis   \n",
       "\n",
       "                                       genome_file_name  \\\n",
       "0                  saccharomyces_cerevisiae_genomic.fna   \n",
       "1                        fusarium_oxysporum_genomic.fna   \n",
       "2                          candida_albicans_genomic.fna   \n",
       "3                   cryptococcus_neoformans_genomic.fna   \n",
       "4                saccharomyces_kudriavzevii_genomic.fna   \n",
       "...                                                 ...   \n",
       "2429  achaetomium_macrosporum_cbs_532_94_v1_0_genomi...   \n",
       "2430         acarospora_strigata_cbs_132363_genomic.fna   \n",
       "2431         absidia_padenii_nrrl_2977_v1_0_genomic.fna   \n",
       "2432         absidia_caerulea_nrrl1315_v1_0_genomic.fna   \n",
       "2433  abortiporus_biennis_cirm_brfm_1778_v1_0_genomi...   \n",
       "\n",
       "                                        cds_file_name  \\\n",
       "0                    saccharomyces_cerevisiae_cds.fna   \n",
       "1                          fusarium_oxysporum_cds.fna   \n",
       "2                            candida_albicans_cds.fna   \n",
       "3                     cryptococcus_neoformans_cds.fna   \n",
       "4                  saccharomyces_kudriavzevii_cds.fna   \n",
       "...                                               ...   \n",
       "2429  achaetomium_macrosporum_cbs_532_94_v1_0_cds.fna   \n",
       "2430           acarospora_strigata_cbs_132363_cds.fna   \n",
       "2431           absidia_padenii_nrrl_2977_v1_0_cds.fna   \n",
       "2432           absidia_caerulea_nrrl1315_v1_0_cds.fna   \n",
       "2433  abortiporus_biennis_cirm_brfm_1778_v1_0_cds.fna   \n",
       "\n",
       "                                original_name  \\\n",
       "0                    saccharomyces_cerevisiae   \n",
       "1                          fusarium_oxysporum   \n",
       "2                            candida_albicans   \n",
       "3                     cryptococcus_neoformans   \n",
       "4                  saccharomyces_kudriavzevii   \n",
       "...                                       ...   \n",
       "2429  achaetomium_macrosporum_cbs_532_94_v1_0   \n",
       "2430           acarospora_strigata_cbs_132363   \n",
       "2431           absidia_padenii_nrrl_2977_v1_0   \n",
       "2432           absidia_caerulea_nrrl1315_v1_0   \n",
       "2433  abortiporus_biennis_cirm_brfm_1778_v1_0   \n",
       "\n",
       "                                                cds_url  \\\n",
       "0     curl -H \"Accept: application/zip\" \"https://api...   \n",
       "1     curl -H \"Accept: application/zip\" \"https://api...   \n",
       "2     curl -H \"Accept: application/zip\" \"https://api...   \n",
       "3     curl -H \"Accept: application/zip\" \"https://api...   \n",
       "4     curl -H \"Accept: application/zip\" \"https://api...   \n",
       "...                                                 ...   \n",
       "2429  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...   \n",
       "2430  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...   \n",
       "2431  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...   \n",
       "2432  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...   \n",
       "2433  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...   \n",
       "\n",
       "                                             genome_url  \\\n",
       "0     curl -H \"Accept: application/zip\" \"https://api...   \n",
       "1     curl -H \"Accept: application/zip\" \"https://api...   \n",
       "2     curl -H \"Accept: application/zip\" \"https://api...   \n",
       "3     curl -H \"Accept: application/zip\" \"https://api...   \n",
       "4     curl -H \"Accept: application/zip\" \"https://api...   \n",
       "...                                                 ...   \n",
       "2429  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...   \n",
       "2430  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...   \n",
       "2431  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...   \n",
       "2432  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...   \n",
       "2433  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...   \n",
       "\n",
       "                                           proteome_url    source  \n",
       "0     curl -H \"Accept: application/zip\" \"https://api...      NCBI  \n",
       "1     curl -H \"Accept: application/zip\" \"https://api...      NCBI  \n",
       "2     curl -H \"Accept: application/zip\" \"https://api...      NCBI  \n",
       "3     curl -H \"Accept: application/zip\" \"https://api...      NCBI  \n",
       "4     curl -H \"Accept: application/zip\" \"https://api...      NCBI  \n",
       "...                                                 ...       ...  \n",
       "2429  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...  MycoCosm  \n",
       "2430  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...  MycoCosm  \n",
       "2431  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...  MycoCosm  \n",
       "2432  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...  MycoCosm  \n",
       "2433  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...  MycoCosm  \n",
       "\n",
       "[2434 rows x 8 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv('fourdbs_input_species.csv')\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species_name</th>\n",
       "      <th>genome_file_name</th>\n",
       "      <th>cds_file_name</th>\n",
       "      <th>original_name</th>\n",
       "      <th>cds_url</th>\n",
       "      <th>genome_url</th>\n",
       "      <th>proteome_url</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1553</th>\n",
       "      <td>pycnoporus_sanguineus</td>\n",
       "      <td>pycnoporus_sanguineus_cirm_brfm_1264_v1_0_geno...</td>\n",
       "      <td>pycnoporus_sanguineus_cirm_brfm_1264_v1_0_cds.fna</td>\n",
       "      <td>pycnoporus_sanguineus_cirm_brfm_1264_v1_0</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...</td>\n",
       "      <td>MycoCosm</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               species_name  \\\n",
       "1553  pycnoporus_sanguineus   \n",
       "\n",
       "                                       genome_file_name  \\\n",
       "1553  pycnoporus_sanguineus_cirm_brfm_1264_v1_0_geno...   \n",
       "\n",
       "                                          cds_file_name  \\\n",
       "1553  pycnoporus_sanguineus_cirm_brfm_1264_v1_0_cds.fna   \n",
       "\n",
       "                                  original_name  \\\n",
       "1553  pycnoporus_sanguineus_cirm_brfm_1264_v1_0   \n",
       "\n",
       "                                                cds_url  \\\n",
       "1553  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...   \n",
       "\n",
       "                                             genome_url  \\\n",
       "1553  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...   \n",
       "\n",
       "                                           proteome_url    source  \n",
       "1553  curl -c cookies.txt -d \"login=YOUR_JGI_USERNAM...  MycoCosm  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['species_name'] == 'pycnoporus_sanguineus']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fungi incertae sedis\n",
    "\n",
    "Incertae sedis (Latin for 'of uncertain placement') or problematica is a term used for a taxonomic group where its broader relationships are unknown or undefined. Alternatively, such groups are frequently referred to as \"enigmatic taxa\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cellular organisms; Eukaryota; Opisthokonta; Metazoa; Porifera; Demospongiae; Heteroscleromorpha; Spongillida; Metschnikowiidae\n",
      "cellular organisms; Eukaryota; Opisthokonta; Fungi; Dikarya; Basidiomycota; Agaricomycotina; Agaricomycetes; Agaricomycetes incertae sedis; Polyporales; Polyporaceae; Trametes\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "def lineage_from_id(taxonomy_id):\n",
    "    # URL for efetch, which retrieves taxonomy data given a taxonomy ID\n",
    "    efetch_url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=taxonomy&id={taxonomy_id}\"\n",
    "\n",
    "    # Send a request to the efetch URL\n",
    "    response = requests.get(efetch_url)\n",
    "    # Parse the XML response\n",
    "    tree = ET.ElementTree(ET.fromstring(response.content))\n",
    "\n",
    "    # Find the Lineage element and extract the taxonomic hierarchy\n",
    "    lineage = None\n",
    "    for elem in tree.iterfind('Taxon/Lineage'):\n",
    "        lineage = elem.text\n",
    "\n",
    "    return lineage\n",
    "\n",
    "def get_name(taxonomy_id):\n",
    "    # URL for efetch, which retrieves taxonomy data given a taxonomy ID\n",
    "    efetch_url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=taxonomy&id={taxonomy_id}\"\n",
    "    \n",
    "    # Send a request to the efetch URL\n",
    "    response = requests.get(efetch_url)\n",
    "    # Parse the XML response\n",
    "    tree = ET.ElementTree(ET.fromstring(response.content))\n",
    "\n",
    "    tree.write('file.txt')\n",
    "    print(tree)\n",
    "\n",
    "def get_taxonomic_hierarchy(species_name):\n",
    "    # URL for esearch, which retrieves primary IDs for a particular search term\n",
    "    esearch_url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=taxonomy&term={species_name}&retmode=json\"\n",
    "    \n",
    "    # Send a request to the esearch URL\n",
    "    response = requests.get(esearch_url)\n",
    "    data = response.json()\n",
    "    \n",
    "    # Check if there are any results\n",
    "    if not data[\"esearchresult\"][\"idlist\"]:\n",
    "        # print(\"No exact match found. Trying a fuzzy search...\")\n",
    "        \n",
    "        # Modify the search term for fuzzy search\n",
    "        fuzzy_species_name = species_name.split(' ')[0]  # use only the genus for the fuzzy search\n",
    "        fuzzy_search_url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=taxonomy&term={fuzzy_species_name}&retmode=json\"\n",
    "        \n",
    "        # Send a request for fuzzy search\n",
    "        response = requests.get(fuzzy_search_url)\n",
    "        data = response.json()\n",
    "        \n",
    "        # Check if the fuzzy search returns results\n",
    "        if not data[\"esearchresult\"][\"idlist\"]:\n",
    "            # print(\"No results found with the fuzzy search either.\")\n",
    "            return \"Not found\"\n",
    "        \n",
    "        # If results are found, use the first ID by default\n",
    "        # For a more advanced implementation, you could present the list of matches to the user and let them choose\n",
    "        # print(\"Fuzzy search successful.\")\n",
    "        taxonomy_ids = data[\"esearchresult\"][\"idlist\"]\n",
    "\n",
    "        taxonomy_id = taxonomy_ids[0]\n",
    "    else:\n",
    "        # If an exact match is found, use the ID\n",
    "        taxonomy_id = data[\"esearchresult\"][\"idlist\"][0]\n",
    "\n",
    "    return lineage_from_id(taxonomy_id)\n",
    "\n",
    "# Replace 'Ramularia collo' with your species of interest\n",
    "species_name = \"Metschnikowia sp\"\n",
    "print(get_taxonomic_hierarchy(species_name))\n",
    "\n",
    "print(lineage_from_id(158606))\n",
    "# get_name(747676)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species_name</th>\n",
       "      <th>classification</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>cellular organisms; Eukaryota; Opisthokonta; F...</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fusarium oxysporum</td>\n",
       "      <td>cellular organisms; Eukaryota; Opisthokonta; F...</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Candida albicans</td>\n",
       "      <td>cellular organisms; Eukaryota; Opisthokonta; F...</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cryptococcus neoformans</td>\n",
       "      <td>cellular organisms; Eukaryota; Opisthokonta; F...</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Saccharomyces kudriavzevii</td>\n",
       "      <td>cellular organisms; Eukaryota; Opisthokonta; F...</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2429</th>\n",
       "      <td>Achaetomium macrosporum</td>\n",
       "      <td>cellular organisms; Eukaryota; Opisthokonta; F...</td>\n",
       "      <td>MycoCosm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2430</th>\n",
       "      <td>Acarospora strigata</td>\n",
       "      <td>cellular organisms; Eukaryota; Opisthokonta; F...</td>\n",
       "      <td>MycoCosm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2431</th>\n",
       "      <td>Absidia padenii</td>\n",
       "      <td>cellular organisms; Eukaryota; Opisthokonta; F...</td>\n",
       "      <td>MycoCosm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2432</th>\n",
       "      <td>Absidia caerulea</td>\n",
       "      <td>cellular organisms; Eukaryota; Opisthokonta; F...</td>\n",
       "      <td>MycoCosm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2433</th>\n",
       "      <td>Abortiporus biennis</td>\n",
       "      <td>cellular organisms; Eukaryota; Opisthokonta; F...</td>\n",
       "      <td>MycoCosm</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2434 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    species_name  \\\n",
       "0       Saccharomyces cerevisiae   \n",
       "1             Fusarium oxysporum   \n",
       "2               Candida albicans   \n",
       "3        Cryptococcus neoformans   \n",
       "4     Saccharomyces kudriavzevii   \n",
       "...                          ...   \n",
       "2429     Achaetomium macrosporum   \n",
       "2430         Acarospora strigata   \n",
       "2431             Absidia padenii   \n",
       "2432            Absidia caerulea   \n",
       "2433         Abortiporus biennis   \n",
       "\n",
       "                                         classification    source  \n",
       "0     cellular organisms; Eukaryota; Opisthokonta; F...      NCBI  \n",
       "1     cellular organisms; Eukaryota; Opisthokonta; F...      NCBI  \n",
       "2     cellular organisms; Eukaryota; Opisthokonta; F...      NCBI  \n",
       "3     cellular organisms; Eukaryota; Opisthokonta; F...      NCBI  \n",
       "4     cellular organisms; Eukaryota; Opisthokonta; F...      NCBI  \n",
       "...                                                 ...       ...  \n",
       "2429  cellular organisms; Eukaryota; Opisthokonta; F...  MycoCosm  \n",
       "2430  cellular organisms; Eukaryota; Opisthokonta; F...  MycoCosm  \n",
       "2431  cellular organisms; Eukaryota; Opisthokonta; F...  MycoCosm  \n",
       "2432  cellular organisms; Eukaryota; Opisthokonta; F...  MycoCosm  \n",
       "2433  cellular organisms; Eukaryota; Opisthokonta; F...  MycoCosm  \n",
       "\n",
       "[2434 rows x 3 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl = pd.read_csv('species_w_classifications.csv')\n",
    "\n",
    "cl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "txt = '''Fungi\n",
    "|-- Chytridiomycota\n",
    "|-- Zoopagomycota\n",
    "|-- Mucoromycota\n",
    "|-- Dikarya\n",
    "|   |-- Basidiomycota\n",
    "|   |   |-- Pucciniomycotina\n",
    "|   |   |-- Ustilaginomycotina\n",
    "|   |   |-- Agaricomycotina\n",
    "|   |   |   |-- Agaricomycetes\n",
    "|   |   |   |-- Dacrymycetes\n",
    "|   |   |   |-- Tremellomycetes\n",
    "|   |   |   |-- Wallemiomycetes\n",
    "|   |-- Ascomycota\n",
    "|       |-- Pezizomycotina\n",
    "|       |   |-- Pezizomycetes\n",
    "|       |   |-- Orbiliomycetes\n",
    "|       |   |-- Eurotiomycetes\n",
    "|       |   |-- Dothideomycetes\n",
    "|       |   |-- Lecanoromycetes\n",
    "|       |   |-- Leotiomycetes\n",
    "|       |   |-- Sordariomycetes\n",
    "|       |   |-- Xylonomycetes\n",
    "|       |-- Saccharomycotina\n",
    "|       |-- Taphrinomycotina\n",
    "|-- Mucoromycotina\n",
    "|       |-- Glomeromycotina\n",
    "|       |-- Mortierellomycotina\n",
    "|       |-- Mucoromycotina\n",
    "|-- Zoopagomycota\n",
    "|       |-- Zoopagomycotina\n",
    "|       |-- Entomophthoromycotina\n",
    "|       |-- Kickxellomycotina\n",
    "|-- Blastocladiomycota\n",
    "|-- Chytridiomycota\n",
    "|       |-- Chytridiomycetes\n",
    "|       |-- Monoblepharidomycetes\n",
    "|       |-- Neocallimastigomycetes\n",
    "|-- Microsporidia\n",
    "|-- Cryptomycota\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "cl = pd.read_csv('species_w_classifications.csv')\n",
    "\n",
    "# Total: 2434\n",
    "fungi = {\n",
    "    \"Chytridiomycota\": [],  # 58\n",
    "\n",
    "    \"Dikarya\": {  # 2211\n",
    "        \"Basidiomycota\": {  # 618\n",
    "            \"Pucciniomycotina\": [],  # 56\n",
    "            \"Ustilaginomycotina\": [],  # 57\n",
    "            \"Wallemiomycotina\": [],  # 3\n",
    "            \"Agaricomycotina\": {  # 502\n",
    "                \"Geminibasidiomycetes\": [],  # 1\n",
    "                \"Agaricomycetes\": [],  # 447\n",
    "                \"Dacrymycetes\": [],  # 9\n",
    "                \"Tremellomycetes\": []  # 45\n",
    "            }\n",
    "        },\n",
    "        \"Ascomycota\": {  # 1593\n",
    "            \"Pezizomycotina\": {  # 1399\n",
    "                \"Pezizomycotina incertae sedis\": [],  # 2 # uncertain placement\n",
    "                \"Pezizomycetes\": [],  # 73\n",
    "                \"Orbiliomycetes\": [],  # 10\n",
    "                \"Leotiomyceta\": {  # 1314\n",
    "                    \"Geoglossomycetes\": [],  # 1\n",
    "                    \"Candelariomycetes\": [], # 2\n",
    "                    \"Coniocybomycetes\": [],  # 3\n",
    "                    \"Eurotiomycetes\": [],  # 417\n",
    "                    \"Dothideomycetes\": [],  # 232\n",
    "                    \"Lecanoromycetes\": [],  # 39\n",
    "                    \"Lichinomycetes\": [],  # 1\n",
    "                    \"Leotiomycetes\": [],  # 108\n",
    "                    \"Sordariomycetes\": [],  # 509\n",
    "                    \"Xylonomycetes\": []  # 2\n",
    "                }\n",
    "            },\n",
    "            \"Saccharomycotina\": [],  # 180\n",
    "            \"Taphrinomycotina\": [],  # 14\n",
    "        }\n",
    "    },\n",
    "    \"Mucoromycota\": {  # 125\n",
    "        \"Mucoromycota incertae sedis\": [],  # 1\n",
    "        \"Glomeromycotina\": [],  # 11\n",
    "        \"Mortierellomycotina\": [],  # 44\n",
    "        \"Mucoromycotina\": {  # 69\n",
    "            \"Umbelopsidomycetes\": [],  # 5\n",
    "            \"Mucoromycetes\": [],  # 60\n",
    "            \"Endogonomycetes\": [],  # 4\n",
    "        },\n",
    "    },\n",
    "    \"Zoopagomycota\": {  # 32\n",
    "        \"Zoopagomycotina\": [],  # 6\n",
    "        \"Entomophthoromycotina\": [],  # 7\n",
    "        \"Kickxellomycotina\": []  # 19\n",
    "    },\n",
    "    \"Olpidiomycota\": [],  # 1\n",
    "    \"Blastocladiomycota\": [],  # 5\n",
    "    \"Chytridiomycota\": {  # 58\n",
    "        \"Chytridiomycetes\": [],  # 45\n",
    "        \"Monoblepharidomycetes\": [],  # 3\n",
    "        \"Neocallimastigomycetes\": []  # 10\n",
    "    },\n",
    "    \"Cryptomycota\": [],  # 1\n",
    "    \"Unclassified Fungi\": []  # 1\n",
    "}\n",
    "\n",
    "unplaced = list()\n",
    "\n",
    "c = 0\n",
    "for idx, row in cl.iterrows():\n",
    "    hierarchy = row['classification']\n",
    "    hierarchy_list = hierarchy.split(\"; \")\n",
    "\n",
    "    if hierarchy_list[0] != 'cellular organisms':\n",
    "        print(\"Not a cellular organism??\")\n",
    "        print(row['species_name'])\n",
    "    if hierarchy_list[1] != 'Eukaryota':\n",
    "        print(\"Not a Eukaryote??\")\n",
    "        print(row['species_name'])\n",
    "    if hierarchy_list[2] != 'Opisthokonta':\n",
    "        print('Not an Opisthokonta??')\n",
    "        print(row['species_name'])\n",
    "    if hierarchy_list[3] != 'Fungi':\n",
    "        print(\"Not a FUNGI??\")\n",
    "        print(row['species_name'])\n",
    "\n",
    "    placed_somewhere = False\n",
    "\n",
    "    current_dict = fungi\n",
    "    for item in hierarchy_list[4:-1]:  # Exclude the last item (Abortiporus) as it's the name we'll be inserting\n",
    "        item = item[0].upper() + item[1:]\n",
    "        if item not in current_dict:\n",
    "            continue\n",
    "        else:\n",
    "            current_dict = current_dict[item]\n",
    "\n",
    "            if isinstance(current_dict, list):\n",
    "                current_dict.append(row['species_name'])\n",
    "                placed_somewhere = True\n",
    "                break\n",
    "\n",
    "    if not placed_somewhere:\n",
    "        unplaced.append(row['species_name'])\n",
    "    else:\n",
    "        c += 1\n",
    "        \n",
    "unplaced"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Fungi': 2434,\n",
       " 'Chytridiomycota': 58,\n",
       " 'Chytridiomycetes': 45,\n",
       " 'Monoblepharidomycetes': 3,\n",
       " 'Neocallimastigomycetes': 10,\n",
       " 'Dikarya': 2211,\n",
       " 'Basidiomycota': 618,\n",
       " 'Pucciniomycotina': 56,\n",
       " 'Ustilaginomycotina': 57,\n",
       " 'Wallemiomycotina': 3,\n",
       " 'Agaricomycotina': 502,\n",
       " 'Geminibasidiomycetes': 1,\n",
       " 'Agaricomycetes': 447,\n",
       " 'Dacrymycetes': 9,\n",
       " 'Tremellomycetes': 45,\n",
       " 'Ascomycota': 1593,\n",
       " 'Pezizomycotina': 1399,\n",
       " 'Pezizomycotina incertae sedis': 2,\n",
       " 'Pezizomycetes': 73,\n",
       " 'Orbiliomycetes': 10,\n",
       " 'Leotiomyceta': 1314,\n",
       " 'Geoglossomycetes': 1,\n",
       " 'Candelariomycetes': 2,\n",
       " 'Coniocybomycetes': 3,\n",
       " 'Eurotiomycetes': 417,\n",
       " 'Dothideomycetes': 232,\n",
       " 'Lecanoromycetes': 39,\n",
       " 'Lichinomycetes': 1,\n",
       " 'Leotiomycetes': 108,\n",
       " 'Sordariomycetes': 509,\n",
       " 'Xylonomycetes': 2,\n",
       " 'Saccharomycotina': 180,\n",
       " 'Taphrinomycotina': 14,\n",
       " 'Mucoromycota': 125,\n",
       " 'Mucoromycota incertae sedis': 1,\n",
       " 'Glomeromycotina': 11,\n",
       " 'Mortierellomycotina': 44,\n",
       " 'Mucoromycotina': 69,\n",
       " 'Umbelopsidomycetes': 5,\n",
       " 'Mucoromycetes': 60,\n",
       " 'Endogonomycetes': 4,\n",
       " 'Zoopagomycota': 32,\n",
       " 'Zoopagomycotina': 6,\n",
       " 'Entomophthoromycotina': 7,\n",
       " 'Kickxellomycotina': 19,\n",
       " 'Olpidiomycota': 1,\n",
       " 'Blastocladiomycota': 5,\n",
       " 'Cryptomycota': 1,\n",
       " 'Unclassified Fungi': 1}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_dict = dict()\n",
    "def recursive_element_counts(level, current_dict, counts_dict):\n",
    "    if level not in counts_dict:\n",
    "        counts_dict[level] = 0\n",
    "\n",
    "    if isinstance(current_dict, dict):\n",
    "        for key in current_dict.keys():\n",
    "            counts_dict = recursive_element_counts(key, current_dict[key], counts_dict)\n",
    "            counts_dict[level] += counts_dict[key]\n",
    "    else:\n",
    "        counts_dict[level] = len(current_dict)\n",
    "        return counts_dict\n",
    "    \n",
    "    return counts_dict\n",
    "\n",
    "recursive_element_counts('Fungi', fungi, counts_dict)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "allegro",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
